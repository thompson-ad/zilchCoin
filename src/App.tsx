import * as React from 'react';
import {View, StyleSheet, ActivityIndicator, Text} from 'react-native';
import StyleGuide from './theme/StyleGuide';
import Button from './components/Button';
import {CardInterface, fetchAllCards, cashInCoins, AsyncStatus} from './api';
import CardSlider from './components/CardSlider';

const App = () => {
  const [cards, setCards] = React.useState<CardInterface[]>([]);
  const [currentCard, setCurrentCard] = React.useState<CardInterface | null>(
    null,
  );
  const [fetchCardsAsyncStatus, setFetchCardsAsyncStatus] =
    React.useState<AsyncStatus>('IDLE');

  const [cashInCoinsAsyncStatus, setcashInCoinsAsyncStatus] =
    React.useState<AsyncStatus>('IDLE');

  React.useEffect(() => {
    setFetchCardsAsyncStatus('PENDING');
    fetchAllCards()
      .then(cardData => {
        setCards(cardData);
        setCurrentCard(cardData[0]);
        // fake lagging connection
        setTimeout(() => {
          setFetchCardsAsyncStatus('RESOlVED');
        }, 1000);
      })
      .catch(error => {
        setFetchCardsAsyncStatus('REJECTED');
        console.error(error);
      });
  }, []);

  const handleOnPress = () => {
    if (!currentCard) {
      return;
    }
    setcashInCoinsAsyncStatus('PENDING');
    cashInCoins(currentCard.id)
      .then(newCardData => {
        const newCards = cards.map(card => {
          if (card.id === currentCard.id) {
            return newCardData;
          } else {
            return card;
          }
        });
        setCards(newCards);
        // fake lagging connection
        setTimeout(() => {
          setcashInCoinsAsyncStatus('RESOlVED');
        }, 1000);
      })
      .catch(error => {
        setcashInCoinsAsyncStatus('REJECTED');
        console.error(error);
      });
  };

  if (fetchCardsAsyncStatus === 'PENDING') {
    return <ActivityIndicator style={styles.activityIndicator} />;
  }

  if (fetchCardsAsyncStatus !== 'IDLE' && !cards.length) {
    return (
      <View style={styles.reminderScreen}>
        <Text style={styles.reminderText}>
          Don't forget to start the server!
        </Text>
      </View>
    );
  }

  return (
    <View style={styles.container}>
      {cards.length > 0 && (
        <CardSlider
          cards={cards}
          setCurrentCard={setCurrentCard}
          asyncStatus={cashInCoinsAsyncStatus}
        />
      )}
      <Button
        text={'Cash in card!'}
        isLoading={cashInCoinsAsyncStatus === 'PENDING'}
        onPress={handleOnPress}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: StyleGuide.palette.backgroundPrimary,
  },
  activityIndicator: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: StyleGuide.palette.backgroundPrimary,
  },
  reminderText: {
    fontSize: 40,
    textAlign: 'center',
  },
  reminderScreen: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export default App;
