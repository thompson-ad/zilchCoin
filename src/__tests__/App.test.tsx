import fetchMock, {enableFetchMocks} from 'jest-fetch-mock';
import React from 'react';
import {render, waitFor, fireEvent} from 'react-native-testing-library';
import App from '../App';
enableFetchMocks();

beforeEach(() => {
  fetchMock.resetMocks();
});

// Snapshot test should ideally be the last
// tests that you write as they will fail a lot
test('it renders as expected', () => {
  const {toJSON} = render(<App />);
  expect(toJSON()).toMatchSnapshot();
});

describe('App Integration Test', () => {
  test('Cards are fetched and displayed', async () => {
    fetchMock.mockResponses([
      JSON.stringify([
        {
          zilchCoinCount: 0,
          id: 0,
        },
        {
          zilchCoinCount: 0,
          id: 1,
        },
        {
          zilchCoinCount: 0,
          id: 2,
        },
      ]),
      {status: 200},
    ]);

    const page = render(<App />);
    const cards = await page.findAllByTestId('card');
    await waitFor(() => {
      expect(cards.length).toEqual(3);
    });
  });

  test('Coin count is reduced to zero when button is pressed', async () => {
    fetchMock.mockResponses(
      [
        JSON.stringify([
          {
            zilchCoinCount: 0,
            id: 0,
          },
          {
            zilchCoinCount: 0,
            id: 1,
          },
          {
            zilchCoinCount: 0,
            id: 2,
          },
        ]),
        {status: 200},
      ],
      [
        JSON.stringify({
          zilchCoinCount: 0,
          id: 0,
        }),
        {status: 200},
      ],
    );

    const page = render(<App />);
    const button = await page.findByTestId('button');
    fireEvent.press(button);
    const coinCountArray = await page.findAllByTestId('coinCount');
    await waitFor(() => {
      expect(coinCountArray[0].props.children).toEqual(0);
    });
  });
});
