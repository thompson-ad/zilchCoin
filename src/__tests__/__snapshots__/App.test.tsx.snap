// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders as expected 1`] = `
<ActivityIndicator
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FDFFFC",
      "flex": 1,
      "justifyContent": "center",
    }
  }
/>
`;
